[profiler] add RecordFunctionAsync struct

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33030 [profiler] add RecordFunctionAsync struct**
getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
We previously had an issue (described in #32517, #32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")`...)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As discussed in this PR, the proposal is to add a new `RecordFunctionAsync` that exposes three different APIs:

`before` --> starts the profiling, and also calls into `RecordFunction::setThreadId`
`exitScope` --> resets the thread local state, needed for this to work with nested profiling
`end` --> invokes the end callbacks (different from `RecordFunction::end()`, which invokes the end callbacks and does the logic of `exitScope` in one function.

The next PR will change the RPC layer to use this and remove the unneeded changes in `RecordFunction`.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

rohan-varma committed:
[profiler] for async RPC, use separate api in RecordFunction to handle nested scopes.

WIP reference implementation, allows nested scopes to be correctly
handled in the case of nested scopes for the profiler.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler] for async RPC, use separate api in RecordFunction to handle nested scopes."


We previously had an issue (described in https://github.com/pytorch/pytorch/issues/32517, https://github.com/pytorch/pytorch/issues/32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")...`)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As proposed in https://github.com/pytorch/pytorch/issues/32884, this change adds 2 APIs to RecordFunction:
`runEndCallbacks()` which just runs the end callbacks for profiling, and doesn't reset thread local RecordFunction data
`resetThreadLocalFunc` that resets the thread local RecordFunction data.

For profiling RPCs, we call `resetThreadLocalFunc` in the creating thread to avoid the above issue.
tested by modifying unit test to add the nested scope.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler] for async RPC, use separate api in RecordFunction to handle nested scopes."


We previously had an issue (described in https://github.com/pytorch/pytorch/issues/32517, https://github.com/pytorch/pytorch/issues/32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")...`)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As proposed in https://github.com/pytorch/pytorch/issues/32884, this change adds 2 APIs to RecordFunction:
`runEndCallbacks()` which just runs the end callbacks for profiling, and doesn't reset thread local RecordFunction data
`resetThreadLocalFunc` that resets the thread local RecordFunction data.

For profiling RPCs, we call `resetThreadLocalFunc` in the creating thread to avoid the above issue.
tested by modifying unit test to add the nested scope.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler] for async RPC, use separate api in RecordFunction to handle nested scopes."


We previously had an issue (described in https://github.com/pytorch/pytorch/issues/32517, https://github.com/pytorch/pytorch/issues/32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")...`)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As proposed in https://github.com/pytorch/pytorch/issues/32884, this change adds 2 APIs to RecordFunction:
`runEndCallbacks()` which just runs the end callbacks for profiling, and doesn't reset thread local RecordFunction data
`resetThreadLocalFunc` that resets the thread local RecordFunction data.

For profiling RPCs, we call `resetThreadLocalFunc` in the creating thread to avoid the above issue.
tested by modifying unit test to add the nested scope.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/autograd/utils.cpp:
```diff
@@ -123,21 +123,27 @@ std::shared_ptr<FutureMessage> sendMessageWithAutograd(
-    // can be profiled correctly.
+    // This must be set synchronously in the same thread to avoid issues with
+    // nested scopes.
+    rf->resetThreadLocalFunc();
```
I moved `rf->setThreadId()` and `rf->resetThreadLocalFunc()` to be before creating the future, since if we do this after, there is a small chance that the future could actually complete before we set the necessary information on the record function and return back to the user.

rohan-varma committed:
Update on "[profiler] for async RPC, use separate api in RecordFunction to handle nested scopes."


We previously had an issue (described in https://github.com/pytorch/pytorch/issues/32517, https://github.com/pytorch/pytorch/issues/32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")...`)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As proposed in https://github.com/pytorch/pytorch/issues/32884, this change adds 2 APIs to RecordFunction:
`runEndCallbacks()` which just runs the end callbacks for profiling, and doesn't reset thread local RecordFunction data
`resetThreadLocalFunc` that resets the thread local RecordFunction data.

For profiling RPCs, we call `resetThreadLocalFunc` in the creating thread to avoid the above issue.
tested by modifying unit test to add the nested scope.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

d4l3k commented:
I spent some time thinking about what the idea API would be for this and I think exposing internal state via methods like resetThreadLocalFunc is the wrong approach here. If we want to have two pretty fundamentally different code paths we should describe it at a higher level. I.e. keep RecordFunction as it is currently and createa a new RecordFunctionAsync which doesn't track the parents at all. The proposed changes requires anyone profiling async code to have a deep understanding of how RecordFunction works internally to avoid writing unsafe code.

The other thought I had was that for tracing purposes which is somewhat orthogonal to async profiling is that we really want to be able to tag metadata on the record functions and not just identify them after the fact. I hacked together some stuff to do the latter approach but it requires a lock on every call which is quite expensive. 

I think the ideal interface for tracing would make RecordFunction provide functionality similar to folly::RequestContext and propagate it all the way through to the backwards pass as well. That way we don't have to use sequence number for correlations and can instead directly attach a trace to the methods.

pritamdamania87 commented:
> I spent some time thinking about what the idea API would be for this and I think exposing internal state via methods like resetThreadLocalFunc is the wrong approach here. If we want to have two pretty fundamentally different code paths we should describe it at a higher level. I.e. keep RecordFunction as it is currently and createa a new RecordFunctionAsync which doesn't track the parents at all. The proposed changes requires anyone profiling async code to have a deep understanding of how RecordFunction works internally to avoid writing unsafe code.

I see your point here, although if RecordFunctionAsync doesn't track parents, it seems like we can't support something like (as mentioned in https://github.com/pytorch/pytorch/issues/32517):

```
with torch.autograd.profiler.profiler() as prof:
    with torch.autograd.profiler.record_function("foo"):
        rpc.rpc_async(...)
```

From a user API perspective, it doesn't seem right to break this functionality.

rohan-varma commented:
@d4l3k @pritamdamania87 given this function, how about something like `RecordFunctionAsync` which essentially just wraps all of the logic that we needed to do here to be able to record parents, so that users adding async profiling later can simply use this class similar to how we expect them to use `RecordFunction`?

d4l3k commented:
I would like to be able to support the use case of arbitrary nested spans etc for tracing purposes. Thread local pointers such as how RecordFunction currently works is very fragile. We could easily extend RecordFunction to support async use cases if we instead switched it to use shared_ptr to point to the parent RF. That way we get memory safety and cross thread compatibility with the trade off of having slightly more overhead.

A hybrid approach might be best

rohan-varma committed:
Update on "[profiler] for async RPC, use separate api in RecordFunction to handle nested scopes."


We previously had an issue (described in https://github.com/pytorch/pytorch/issues/32517, https://github.com/pytorch/pytorch/issues/32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")...`)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As proposed in https://github.com/pytorch/pytorch/issues/32884, this change adds 2 APIs to RecordFunction:
`runEndCallbacks()` which just runs the end callbacks for profiling, and doesn't reset thread local RecordFunction data
`resetThreadLocalFunc` that resets the thread local RecordFunction data.

For profiling RPCs, we call `resetThreadLocalFunc` in the creating thread to avoid the above issue.
tested by modifying unit test to add the nested scope.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma commented:
I discussed with with @wanchaol, @pritamdamania87, and @d4l3k to figure out how we can unblock here and we came up with the following suggestions. 

At a high level, the goal here is to make RPC calls work with the nested scopes in the profiler, they currently don't due to the `thread_local` state in the profiler's `RecordFunction` class. Ideally we would like to remove the `thread_local`s, or we could also work around them (as I've attempted to do in this PR):

1.  Currently, `profiler::record_function_enter` and `profiler::record_function_exit` are registered as torch ops and are therefore profiled. It doesn't seem too useful to profile these, so we can either: 
        a) not register them as torch ops or 
        b) try to disable the `RECORD_FUNCTION` profiling that comes with torch ops, for this op only. 
However, these suggestions won't work since:
        a) they need to be registered as torch ops since this will eventually be called as part of torchscript, and 
        b) We shouldn't really add all the hooks needed for that just for this one special case. 

2. Create a separate API called `resetThreadLocalFunc()` as proposed by @pritamdamania87 and call it when we are creating the future to return to the user. This will ensure all the state in `RecordFunction` is set correctly. However, this is quite fragile.

3. Implement `RecordFunctionAsync` where the `parent` pointer is an `std::shared_ptr`. Although, I'm not sure how this will allow us to work around the issue of `RecordFunction::current()` still being thread-local? cc @d4l3k 

Open to any further discussion on this.

pritamdamania87 commented:
@rohan-varma I would prefer option 3 here just to have a cleaner separate async API. Regarding the thread local, you are right I don't think we can get rid of the thread local and the RecordFunctionAsync APIs need to update the thread local variable appropriately. For example if could have 3 APIs:

1. before() -> start profiling the async op.
2. exit_scope() -> reset the thread local to the parent to ensure `with record_function` works correctly. We could have a guard that calls exit_scope() appropriately when needed.
3. end() -> end profiling for the async op.

rohan-varma commented:
@pritamdamania87 Sounds good, I will update this PR to implement RecordFunctionAsync.

rohan-varma committed:
Update on "[profiler] for async RPC, use separate api in RecordFunction to handle nested scopes."


We previously had an issue (described in https://github.com/pytorch/pytorch/issues/32517, https://github.com/pytorch/pytorch/issues/32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")...`)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As proposed in https://github.com/pytorch/pytorch/issues/32884, this change adds 2 APIs to RecordFunction:
`runEndCallbacks()` which just runs the end callbacks for profiling, and doesn't reset thread local RecordFunction data
`resetThreadLocalFunc` that resets the thread local RecordFunction data.

For profiling RPCs, we call `resetThreadLocalFunc` in the creating thread to avoid the above issue.
tested by modifying unit test to add the nested scope.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma committed:
Update on " [profiler] add RecordFunctionAsync struct"

functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct.**

We previously had an issue (described in #32517, #32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")`...)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As discussed in this PR, the proposal is to add a new `RecordFunctionAsync` that exposes three different APIs:

`before` --> starts the profiling, and also calls into `RecordFunction::setThreadId`
`exitScope` --> resets the thread local state, needed for this to work with nested profiling
`end` --> invokes the end callbacks (different from `RecordFunction::end()`, which invokes the end callbacks and does the logic of `exitScope` in one function.

The next PR will change the RPC layer to use this and remove the unneeded changes in `RecordFunction`.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma committed:
Update on " [profiler] add RecordFunctionAsync struct"

functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct.**

We previously had an issue (described in #32517, #32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")`...)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As discussed in this PR, the proposal is to add a new `RecordFunctionAsync` that exposes three different APIs:

`before` --> starts the profiling, and also calls into `RecordFunction::setThreadId`
`exitScope` --> resets the thread local state, needed for this to work with nested profiling
`end` --> invokes the end callbacks (different from `RecordFunction::end()`, which invokes the end callbacks and does the logic of `exitScope` in one function.

The next PR will change the RPC layer to use this and remove the unneeded changes in `RecordFunction`.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma committed:
Update on " [profiler] add RecordFunctionAsync struct"

getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct.**

We previously had an issue (described in #32517, #32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")`...)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As discussed in this PR, the proposal is to add a new `RecordFunctionAsync` that exposes three different APIs:

`before` --> starts the profiling, and also calls into `RecordFunction::setThreadId`
`exitScope` --> resets the thread local state, needed for this to work with nested profiling
`end` --> invokes the end callbacks (different from `RecordFunction::end()`, which invokes the end callbacks and does the logic of `exitScope` in one function.

The next PR will change the RPC layer to use this and remove the unneeded changes in `RecordFunction`.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma committed:
Update on " [profiler] add RecordFunctionAsync struct"

getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct.**

We previously had an issue (described in #32517, #32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")`...)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As discussed in this PR, the proposal is to add a new `RecordFunctionAsync` that exposes three different APIs:

`before` --> starts the profiling, and also calls into `RecordFunction::setThreadId`
`exitScope` --> resets the thread local state, needed for this to work with nested profiling
`end` --> invokes the end callbacks (different from `RecordFunction::end()`, which invokes the end callbacks and does the logic of `exitScope` in one function.

The next PR will change the RPC layer to use this and remove the unneeded changes in `RecordFunction`.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

pritamdamania87 commented on test/test_autograd.py:
```diff
@@ -2600,9 +2605,54 @@ def test_record_function_callbacks(self):
+
+        self.verify_function_recorded(p.function_events, "foo")
+
+    def test_record_function_async(self):
```
Nice unit tests!

pritamdamania87 commented on torch/csrc/autograd/init.cpp:
```diff
@@ -59,6 +59,19 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject *unused) {
+            recordFunctionAsync.before(std::move(name));
+          })
+      .def("exit_scope", &RecordFunctionAsync::exitScope)
+      .def("end", &RecordFunctionAsync::end);
```
Should we use `py::gil_scoped_release` for these?

pritamdamania87 commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -242,6 +242,53 @@ RecordFunction* RecordFunction::current() {
+
+void RecordFunctionAsync::before(Node* fn, int64_t sequence_nr) {
+  RecordFunction::before(fn, sequence_nr);
+   RecordFunction::setThreadId();
```
nit: indentation.

pritamdamania87 commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -242,6 +242,53 @@ RecordFunction* RecordFunction::current() {
+      }
+    }
+    initialized_ = false;
+  }
```
nit: Maybe split this part of RecordFunction::end() into a separate protected helper method in RecordFunction and just call that here.

rohan-varma replied:
I'll add a helper for the part that runs the ending callbacks, although RecordFunction::end() has some more logic before it sets initialized = false.

pritamdamania87 commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -242,6 +242,53 @@ RecordFunction* RecordFunction::current() {
+
+// toggle initialized so that RecordFunction destructor does not attempt to
+// reinvoke callbacks.
+RecordFunctionAsync::~RecordFunctionAsync() {
```
If `exitScope` and `end` haven't been called yet, this should be an error here right? Can we add a `DCHECK` here ensuring both `exitScope` and `end` have been called?

rohan-varma replied:
I guess it is a logic error, though technically, if the user just creates and destroys a `RecordFunctionAsync`, nothing will happen, since:
- before() is not called, so starting callbacks are not run
- end() is not called, so ending callbacks are not run, and they are not run in the RecordFn destructor either, since we reset initialized_ = false.

If we decide to check, is the best way to do this to add booleans indicating the functions have been called?

rohan-varma replied:
Also `RecordFunction` doesn't currently contain error handling if `before()` is not called at all or called after `End`, maybe we should check this also if we decide to go this route.

rohan-varma replied:
I added some error handling in the top diff in this stack.

pritamdamania87 replied:
If thread_local_func == this, it means exit_scope hasn't been called, if initialized_ = true, it means end() hasn't been called. We can raise errors for these cases.

pritamdamania87 replied:
> Also RecordFunction doesn't currently contain error handling if before() is not called at all or called after End, maybe we should check this also if we decide to go this route.

Might be fine for RecordFunction since its single threaded and the RECORD_FUNCTION guard takes care of most things.

rohan-varma replied:
> If thread_local_func == this, it means exit_scope hasn't been called, if initialized_ = true, it means end() hasn't been called. We can raise errors for these cases.

Should we log them or actually call `TORCH_INTERNAL_ASSERT` that will throw an exception? This is a destructor and I thought it was generally bad to throw in the destructor. Also, I think clang-tidy would complain if we did.

rohan-varma committed:
Update on " [profiler] add RecordFunctionAsync struct"

getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct.**

We previously had an issue (described in #32517, #32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")`...)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As discussed in this PR, the proposal is to add a new `RecordFunctionAsync` that exposes three different APIs:

`before` --> starts the profiling, and also calls into `RecordFunction::setThreadId`
`exitScope` --> resets the thread local state, needed for this to work with nested profiling
`end` --> invokes the end callbacks (different from `RecordFunction::end()`, which invokes the end callbacks and does the logic of `exitScope` in one function.

The next PR will change the RPC layer to use this and remove the unneeded changes in `RecordFunction`.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma committed:
Update on " [profiler] add RecordFunctionAsync struct"

getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**

We previously had an issue (described in #32517, #32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")`...)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As discussed in this PR, the proposal is to add a new `RecordFunctionAsync` that exposes three different APIs:

`before` --> starts the profiling, and also calls into `RecordFunction::setThreadId`
`exitScope` --> resets the thread local state, needed for this to work with nested profiling
`end` --> invokes the end callbacks (different from `RecordFunction::end()`, which invokes the end callbacks and does the logic of `exitScope` in one function.

The next PR will change the RPC layer to use this and remove the unneeded changes in `RecordFunction`.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/autograd/record_function.h:
```diff
@@ -121,8 +121,11 @@ struct TORCH_API RecordFunction {
   void processCallbacks();
+  // Runs the end callbacks that were pushed to the callback manager. Throws if
+  // the current RecordFunction is not initialized.
+  void runEndCallbacks();
```
Looks like this function and `processCallbacks` works in as a pair. If this is true, should we name them accordingly, e.g., `runStartCallbacks()` and `runEndCallbacks()`?

rohan-varma replied:
Yes, I think it should be more aptly named to runStartCallbacks. It only invokes the starting ones

rohan-varma committed:
Update on " [profiler] add RecordFunctionAsync struct"

getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
We previously had an issue (described in #32517, #32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")`...)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As discussed in this PR, the proposal is to add a new `RecordFunctionAsync` that exposes three different APIs:

`before` --> starts the profiling, and also calls into `RecordFunction::setThreadId`
`exitScope` --> resets the thread local state, needed for this to work with nested profiling
`end` --> invokes the end callbacks (different from `RecordFunction::end()`, which invokes the end callbacks and does the logic of `exitScope` in one function.

The next PR will change the RPC layer to use this and remove the unneeded changes in `RecordFunction`.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

rohan-varma committed:
Update on " [profiler] add RecordFunctionAsync struct"

getEventList.record()
* #33605 [rpc][profiler] integrate RecordFunctionAsync with RPC, and remove unneeded functions from the general RecordFunction
* **#33030 [profiler] add RecordFunctionAsync struct**
We previously had an issue (described in #32517, #32884) where async RPCs would not be profiled correctly if there were nested scopes during profiling (nested scopes can be created by `with torch.autograd.profiler.record_function("foo")`...)

The cause of this was that the logic for handling nested scopes was not correct when we invoked `RecordFunction::end()` from a separate thread.

As discussed in this PR, the proposal is to add a new `RecordFunctionAsync` that exposes three different APIs:

`before` --> starts the profiling, and also calls into `RecordFunction::setThreadId`
`exitScope` --> resets the thread local state, needed for this to work with nested profiling
`end` --> invokes the end callbacks (different from `RecordFunction::end()`, which invokes the end callbacks and does the logic of `exitScope` in one function.

The next PR will change the RPC layer to use this and remove the unneeded changes in `RecordFunction`.

Differential Revision: [D19739387](https://our.internmc.facebook.com/intern/diff/D19739387/)

[ghstack-poisoned]

ilia-cher commented on torch/csrc/autograd/init.cpp:
```diff
@@ -56,8 +56,29 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject *unused) {
+      .def(
+          "end",
+          &RecordFunctionAsync::end,
+          py::call_guard<py::gil_scoped_release>());
```
as discussed offline, this API is extremely fragile, and we should try hard not to expose internal record functions; let's discuss what of it is really needed to be exposed

ilia-cher reviewed (changes requested):
see comment above
